### PR TITLE
new version of BlindOutputs method

### LINKED ISSUE #6: [blindOutputs] Support blinding only the outputs of wich we have the blinding public key

### PR DESCRIPTION
**Pull request by louisinger**
**This PR adds an optional parameter to `Psbt.BlindOutputs` method: the `outputsIndexToBlind` lets to blind only a selected set of outputs**

- undefined `outputsIndexToBlind` in `BlindOutputs` = blind all the outputs (except fee).
- handle unconfidential inputs.
- write a test with 1 confidential output and 1 unconfidential one in `integration/transaction.spec.ts`.

--> WIP because the test failed at the broadcast step: when I try to manually send the raw transaction, I get the famous and devil "bad-txns-in-ne-out" error: `bad-txns-in-ne-out, value in != value out (code 16)`. I've spent some time to investigate this, I'm pretty sure that it comes from my `proofs` computation but I was not able to find a solution. Any help is welcome :1st_place_medal: 

it closes #6 

@tiero @altafan please review

### CODE REVIEW

File: ts_src/psbt.ts [hunk 1]
```diff
@@ -733,13 +739,12 @@ export class Psbt {
...
 
     // generate output blinding factors
-    const numOutputs = this.data.outputs.length - 1; // exclude fees output
-    const outputAbfs = this.data.outputs
-      .slice(0, numOutputs)
-      .map(() => randomBytes(opts!));
-    const outputVbfs = this.data.outputs
-      .slice(0, numOutputs - 1)
-      .map(() => randomBytes(opts!));
+    const numOutputs = c.__TX.outs.length - 1 // excluding fee output
```
Comment (altafan) [inline ts_src/psbt.ts]: this should not depend from outputs anymore.. it should be:
```suggestion
    const numOutputs = outputsIndexToBlind!.length
```

File: ts_src/psbt.ts [hunk 2]
```diff
@@ -733,13 +739,12 @@ export class Psbt {
...
+    const outputAbfs = range(numOutputs)
+      .map((index: number) => outputsIndexToBlind!.includes(index) ? randomBytes(opts!) : Buffer.alloc(32));
```
Comment (altafan) [inline ts_src/psbt.ts]: at this point, we need to generate the random blinders (so-called blinding factors) for both asset and value of each output that is gonna be blinded. Hence, we could just create outputAbfs (and also outputVbfs) like:
```suggestion
    const outputAbfs = outputsIndexToBlind!.map(() => randomBytes(opts!));
```

File: ts_src/psbt.ts [hunk 3]
```diff
@@ -798,6 +804,14 @@ export class Psbt {
...
     });
 
+    range(numOutputs)
+      .filter(index => !outputsIndexToBlind!.includes(index))
+      .forEach(unconfidentialIndex => {
+        c.__TX.setOutputNonce(unconfidentialIndex, Buffer.alloc(0))
+        // c.__TX.setOutputRangeProof(unconfidentialIndex, Buffer.alloc(0))
+        // c.__TX.setOutputSurjectionProof(unconfidentialIndex, Buffer.alloc(0))
+      })
+
```
Comment (altafan) [inline ts_src/psbt.ts]: what is this?

**Review by tiero — COMMENTED** [review]
I understand the idea behind additional parameter, but would break current software wich are using the lib and become backward incompatible.

Would you rethink of using this without specifiying the indexes you want to blind, rather than checking the address if is confidential or not?

This would change the addOutput interface that will expect the blindin pub key along with the script if no confidential address is provided (being an object adding more optional keys make this backward compatible)

File: ts_src/txutils.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
+import { WitnessUtxo } from 'bip174/src/lib/interfaces';
```
Comment (altafan) [inline ts_src/txutils.ts]: can we remove this? let's try to be consistent with bitcoinjs and avoid adding unnecessary files.

It is just fine to check that the proofs are not null for a confidential output... so `bufferNotEmptyOrNull` and `isConfidentialUtxo` could be removed... we could add a isConfidential method to the Output type of the transaction module, but let's do this in another PR.

tryToUnblind could be moved directly in the pset module but revert to the precedent version.. trying to unblind one prevout with all the blinding keys is too much time expensive in js.

**Review by altafan — CHANGES_REQUESTED** [review]

**Comment by altafan** [comment]
@louisinger to keep backward compatibility it's better to rename the existing `blindOutputs` function that you are changing to `blindOutputsByIndex` and make `blindOutputs` to simply call this new function without specifying an array of indexes. This way, users that already make use of blindOutputs are not affected in any manner by these changes.

File: ts_src/psbt.ts (hunk 3 above)
Comment (louisinger) [reply ts_src/psbt.ts]: I was thinking that it will solve the problem, let's remove it :)

File: ts_src/psbt.ts [hunk 5]
```diff
@@ -659,22 +660,40 @@ export class Psbt {
...
+      outputsIndexToBlind = [];
+      this.__CACHE.__TX.outs.forEach((out: Output, index: number) => {
+        if (out.script.length > 0) outputsIndexToBlind!.push(index);
+      });
+    }
+
+    if (outputsIndexToBlind.length !== blindingPubkeys.length)
       throw new Error(
-        'Blinding public keys do not match the number of outputs (fee excluded)',
+        'not enough blinding public keys to blind all the outputs to blind',
```
Comment (altafan) [inline ts_src/psbt.ts]:
```suggestion
        'not enough blinding public keys to blind the requested outputs',
```

File: ts_src/psbt.ts [hunk 6]
```diff
@@ -687,43 +706,47 @@ export class Psbt {
     const inputAgs: Buffer[] = [];
     const inputValues: string[] = [];
 
+    // counter for fetching blinding private key
+    let confidentialUtxoIndex = 0;
+
     // unblind inputs
     this.data.inputs.forEach((input: PsbtInput, index: number) => {
-      let prevout: any;
+      let prevout: WitnessUtxo;
```
Comment (altafan) [inline ts_src/psbt.ts:763]: it was `any` because of the input.nonWitnessUtxo branch. This assigns to prevout a transaction.Output type rather than WitnessUtxo. Is it safe like that?

File: ts_src/psbt.ts [hunk 7]
```diff
@@ -687,43 +706,47 @@ export class Psbt {
...
       if (input.nonWitnessUtxo) {
         const prevTx = nonWitnessUtxoTxFromCache(c, input, index);
         const prevoutIndex = c.__TX.ins[index].index;
         prevout = prevTx.outs[prevoutIndex];
       } else {
-        prevout = { ...input.witnessUtxo };
+        prevout = input.witnessUtxo!;
```
Comment (altafan) [inline ts_src/psbt.ts]: this was assigning to prevout a copy of the input's witness utxo so that future changes wouldn't have affected directly the partial transaction. I think this might be reverted. Still didn't look at this in deep.

**Comment by louisinger** [comment]
@tiero @altafan I've committed the changes according to your review. Now there is two methods: the former one `BlindOutputs` with the same behavior than the last version (tests passes without modification) and `BlindOutputsIndex` which takes an array of output index to blind. I also remove all the stuff in `txutils` and just move `UnblindWitnessUtxo` in `psbt.ts`. 

However, the tx still invalid.  It seems that the amount is not correctly validated. I'm investigating :thinking: 
Example : https://gist.github.com/louisinger/f9f6d11c5cc936bd0edeaccc7e2d1f6a

File: ts_src/psbt.ts [hunk 8]
```diff
@@ -687,43 +706,47 @@ export class Psbt {
...
-        unblindPrevout.vbf = unblindProof.valueBlindingFactor;
+      let unblindPrevout: UnblindWitnessUtxoResult;
+
+      if (prevout.rangeProof != null && prevout.surjectionProof != null) {
+        const blindingPrivKey = blindingPrivkeys[confidentialUtxoIndex];
+        if (!blindingPrivKey) {
+          throw new Error(
+            'the number of blinding private keys does not match the number of confidential inputs',
+          );
+        }
```
Comment (altafan) [inline ts_src/psbt.ts]: in go-elements, instead of implementing this logic, we opted for having the array of input blinding keys of the same length of the inputs.. in case an input is unconfidential, its key is null. This delegates the user to provide the correct list of blinding keys and the method to just check that in case an input is blinded, its binding key must not be null. 

I would keep the same logic, so for example if my tx has 3 inputs, the first unconf and the others conf, I would need to provide a list of keys like `[null, key1, key2]`.

File: ts_src/psbt.ts [hunk 9]
```diff
@@ -1717,3 +1738,38 @@ function randomBytes(options?: RngOpts): Buffer {
...
+    prevout.asset,
+    prevout.script,
+  );
+
+  unblindPrevout.ag = unblindProof.asset;
+  unblindPrevout.value = unblindProof.value;
+  unblindPrevout.abf = unblindProof.assetBlindingFactor;
+  unblindPrevout.vbf = unblindProof.valueBlindingFactor;
+
+  return unblindPrevout;
```
Comment (altafan) [inline ts_src/psbt.ts]: this could be simplified:
```suggestion
  const unblindProof = confidential.unblindOutput(
    prevout.nonce,
    blindingPrivKey,
    prevout.rangeProof!,
    prevout.value,
    prevout.asset,
    prevout.script,
  );

  return {
    value: unblindProof.value,
    ag: unblindProof.asset,
    abf: unblindProof.assetBlindingFactor,
    vbf: unblindProof.valueBlindingFactor,
  };
```

**Review by altafan — CHANGES_REQUESTED** [review]

File: ts_src/psbt.ts (hunk 6 above)
Comment (louisinger) [reply ts_src/psbt.ts:763]: Yes because `WitnessUtxo` and `Output` are the same interface... but I will add `as WitnessUtxo`

File: ts_src/psbt.ts (hunk 7 above)
Comment (louisinger) [reply ts_src/psbt.ts]: The `prevout` is only used to compute commitments [here](https://github.com/vulpemventures/liquidjs-lib/blob/62d68049b6779a8b873638d82c56048817659a8d/ts_src/psbt.ts#L753-L756). Why future changes would modify the partial ?

File: ts_src/psbt.ts (hunk 8 above)
Comment (louisinger) [reply ts_src/psbt.ts]: OK let's keep the go-elements logic :+1:

File: ts_src/psbt.ts (hunk 7 above)
Comment (tiero) [reply ts_src/psbt.ts]: The future PSET will change the serlization

**Review by tiero — CHANGES_REQUESTED** [review]
blindOutputsIndex should not be used by user IMHO, just as "internal"

Just for the sake of a "easier" API I would add this "higher level" method (that will call the blindsOuptpusIndex under the hood)

 ```
 blindOutputsByIndex(blindingPrivateKeys: Array<Object>, blindingPublicKeys: Array<Object>)
 
 
 eg.
 
 blindOutputsByIndex(
  [
     {
       index: Number,
       blindingPrivateKey: string
     }
  ],
  [
     {
       index: Number,
       blindingPublicKey: string
     }
  ]
)
 ```

File: ts_src/psbt.ts [hunk 10]
```diff
@@ -733,59 +765,57 @@ export class Psbt {
...
-      .slice(0, numOutputs - 1)
-      .map(() => randomBytes(opts!));
+    const numOutputs = outputsIndexToBlind.length;
+    const outputAbfs = range(numOutputs).map(() => randomBytes(opts!));
+    const outputVbfs = range(numOutputs - 1).map(() => randomBytes(opts!));
+
     const finalVbf = confidential.valueBlindingFactor(
       inputValues,
-      outputValues,
+      outputValues.filter((_, index) => outputsIndexToBlind!.includes(index)),
```
Comment (altafan) [inline ts_src/psbt.ts:774]: can you explain this?

**Review by altafan — CHANGES_REQUESTED** [review]

File: ts_src/psbt.ts [hunk 11]
```diff
@@ -659,25 +660,51 @@ export class Psbt {
...
+  ): this {
+    return this.blindOutputsIndex(
+      blindingPrivkeys,
+      blindingPubkeys,
+      undefined,
+      opts,
+    );
+  }
+
+  blindOutputsIndex(
```
[the remark this answers is not in the document]
Comment (tiero) [reply ts_src/psbt.ts]: Read my comment, this is the internal one

File: ts_src/psbt.ts (hunk 10 above)
Comment (tiero) [inline ts_src/psbt.ts]: Better wrap in in variable to make it more readble and to comment in it, I agree

File: ts_src/psbt.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (altafan) [reply ts_src/psbt.ts]: In spike of tiero's comment, let's just add a new blindOutputsByIndex function that calls the existing blindOutputsIndex.

File: src/psbt.js [hunk 12]
```diff
@@ -560,14 +560,31 @@ class Psbt {
...
+      blindingPublicKey.push(pubBlindingKey);
+    }
+    return this.RawBlindOutputs(
+      blindingPrivKeysArgs,
+      blindingPublicKey,
+      outputsIndexToBlind,
+      opts,
+    );
+  }
+  RawBlindOutputs(
```
Comment (tiero) [inline src/psbt.js]: would use lowercase, just to be consistent with the style we have already

**Comment by altafan** [comment]
Remove the yarn.lock file since there's already a package-lock.json.

**Comment by louisinger** [comment]
I've added a test in `transaction.spec.ts` (L600) trying to broadcast a tx with an unblinded input and a blinded one. However, this one does not pass. I'm investigating this...

File: ts_src/psbt.ts [hunk 13]
```diff
@@ -769,16 +769,21 @@ export class Psbt {
...
-      if (unblindPrevout) {
+      const privKey = blindingPrivkeys[index];
+      if (privKey) {
+        const unblindPrevout = getBlindingDataForInput(prevout, privKey);
         inputAgs.push(unblindPrevout.ag);
         inputValues.push(unblindPrevout.value);
         inputAbfs.push(unblindPrevout.abf);
         inputVbfs.push(unblindPrevout.vbf);
+      } else {
+        const zeroBuffer32bytes = Buffer.alloc(32);
```
Comment (tiero) [inline ts_src/psbt.ts]: 🚀

File: ts_src/psbt.ts [hunk 14]
```diff
@@ -769,16 +769,19 @@ export class Psbt {
...
         inputAgs.push(unblindPrevout.ag);
         inputValues.push(unblindPrevout.value);
         inputAbfs.push(unblindPrevout.abf);
         inputVbfs.push(unblindPrevout.vbf);
+      } else {
+        const zeroBuffer32bytes = Buffer.alloc(32)
+        inputValues.push(confidential.confidentialValueToSatoshi(prevout.value).toString(10))
+        inputAgs.push(prevout.asset.slice(1))
+        inputAbfs.push(zeroBuffer32bytes)
+        inputVbfs.push(zeroBuffer32bytes)
```
Comment (altafan) [inline ts_src/psbt.ts]: why not returning this into `getBlindingDataForInpit` in case the private key is undefined? It would increase readability. Also, you could be able to use the ZERO constant of transaction module instead of instantiating this zeroBuffer32Bytes.

File: ts_src/psbt.ts [hunk 15]
```diff
@@ -1759,12 +1762,8 @@ export interface UnblindWitnessUtxoResult {
...
 function getBlindingDataForInput(
   prevout: WitnessUtxo,
-  blindPrivKey?: Buffer,
-): UnblindWitnessUtxoResult | undefined {
-  if (!blindPrivKey) {
-    return undefined;
-  }
-
+  blindPrivKey: Buffer,
+): UnblindWitnessUtxoResult {
```
Comment (altafan) [inline ts_src/psbt.ts]: I don't really like this name because it seems that blinding/unblinding is supported only for segwit/native-segwit outputs this way. I would call this `UnblindPrevoutResult`.. eventually, since this is an internal method, it could simply return `any` given that the mentioned type is just a wrapper of the confidential module's method (we should plan to fix the confidential module so that it returns more "user-friendly"er results... like we did in go-elements basically.

File: test/integration/transaction.spec.ts [hunk 16]
```diff
@@ -37,13 +37,13 @@ describe('liquidjs-lib (transactions with psbt)', () => {
...
       nonWitnessUtxo: Buffer.from(
         '0200000000010caf381d44f094661f2da71a11946251a27d656d6c141577e27c483a6' +
-        'd428f01010000006a47304402205ac99f5988d699d6d9f72004098c2e52c8f342838e' +
-        '9009dde33d204108cc930d022077238cd40a4e4234f1e70ceab8fd6b51c5325954387' +
-        '2e5d9f4bad544918b82ce012102b5214a4f0d6962fe547f0b9cbb241f9df1b61c3c40' +
-        '1dbfb04cdd59efd552bea1ffffffff020125b251070e29ca19043cf33ccd7324e2dda' +
-        'b03ecc4ae0b5e77c4fc0e5cf6c95a010000000005f5df70001976a914659bedb5d3d3' +
-        'c7ab12d7f85323c3a1b6c060efbe88ac0125b251070e29ca19043cf33ccd7324e2dda' +
-        'b03ecc4ae0b5e77c4fc0e5cf6c95a010000000000000190000000000000',
+          'd428f01010000006a47304402205ac99f5988d699d6d9f72004098c2e52c8f342838e' +
```
Comment (altafan) [inline test/integration/transaction.spec.ts]: I don't understand all this kind of changes.... all files should be properly linted and prettiefied already.. all these changes are really annoying :/

File: .gitignore [hunk 17]
```diff
@@ -6,3 +6,5 @@ test/*.js
 test/integration/*.js
 !test/ts-node-register.js
 .DS_Store
+ts_src/*.js
+ts_src/payments/*.js
```
Comment (altafan) [inline .gitignore]: what? remove! in `ts_src` there are no js files. If so, you're doing something really weird.

File: test/integration/transaction.spec.ts [hunk 18]
```diff
@@ -1635,6 +1791,7 @@ async function getInputData(
 
   return {
     hash: Buffer.from(unspent.txid, 'hex').reverse(),
+
```
Comment (altafan) [inline test/integration/transaction.spec.ts]: revert.

File: ts_src/psbt.ts [hunk 19]
```diff
@@ -659,6 +660,64 @@ export class Psbt {
...
+
+  addUnknownKeyValToOutput(outputIndex: number, keyVal: KeyValue): this {
+    this.data.addUnknownKeyValToOutput(outputIndex, keyVal);
+    return this;
+  }
+
+  clearFinalizedInput(inputIndex: number): this {
+    this.data.clearFinalizedInput(inputIndex);
+    return this;
+  }
```
Comment (altafan) [inline ts_src/psbt.ts:714]: can you put these methods where they were? Let's try to keep the existing order.
It's quite hard to review the core changes of the PR since it's plenty of unnecessary changes. Is it possible to polish the code?

File: ts_src/psbt.ts [hunk 20]
```diff
@@ -1717,3 +1754,45 @@ function randomBytes(options?: RngOpts): Buffer {
   const rng = options.rng || _randomBytes;
   return rng(32);
 }
+
+export interface UnblindWitnessUtxoResult {
```
Comment (altafan) [inline ts_src/psbt.ts]: this interface shouldn't be exported.

File: test/integration/transaction.spec.ts (hunk 16 above)
Comment (louisinger) [reply test/integration/transaction.spec.ts]: I only use the command `npm run format && npm run format:ci` that generates this changes. That's mandatory, the CI pipeline fails if format is not apply.

File: .gitignore (hunk 17 above)
Comment (louisinger) [reply .gitignore]: I'm using a vscode extension to debug that generates file in ts_src. That's just for confort :thinking:

Comment (altafan) [reply .gitignore]: This is something that concerns your workspace... I don't have any js file in my ts_src folder, but in general, no js files should be present in there at all.

Please, revert these changes.

**Review by altafan — APPROVED** [review]

**Review by tiero — APPROVED** [review]
